src/app/scopes-dialog.component.ts

```diff
@@ -16,21 +16,89 @@ declare let fabric, mwf;
+    this.scopesListTableHeight = window.getComputedStyle(this.scopesTableList.nativeElement, null).getPropertyValue("height");
   }
 
+  /** 
+   * Indicates whether the scope list has been changed.
+   * @returns {boolean} A value of true indicates that a scope has been added or removed from the scope list.
+  */
   scopeListIsDirty(): boolean {
-    return PermissionScopes.filter((s) => s.enabled !== s.enabledTarget).length > 0;
+
+    // Determine whether the scope list has changed. The scope list has changed if isDirty = true.
+    let isDirty = PermissionScopes.filter((s) => s.enabled !== s.enabledTarget).length > 0;
+
+    // Reduce the size of the scopes table list by the size of the message bar. We only want to make this
+    // change the first time that the selected scope list is changed. 
+    if (isDirty && !this.hasChangedScopeListHeight) {
+
+      // Convert the table height from string to number.
+      var currentHeight: number = Number(this.scopesListTableHeight.replace(/px/, "")).valueOf();
+
+      // Update the table height based on the height of the message bar that's displayed when the scope list is dirty.
+      // This should keep the buttons from moving below the viewport as long as the component is showing in the viewport.
+      var updatedHeight: number = currentHeight - 60;
+
+      // Convert the updated height to a string and set the scopes list table height style.
+      this.scopesTableList.nativeElement.style.height = updatedHeight.toString() + "px";
+
+      // We only want to adjust the height one time after making a change to the selected scope list.
+      this.hasChangedScopeListHeight = true;
+    }
+
+    return isDirty;
   }
 
+  /** 
+   * Indicates whether an admin scope has been selected.
+   * @returns {boolean} A value of true indicates that a scope that requires admin consent has been selected.
+  */
   requestingAdminScopes(): boolean {
-    return PermissionScopes.filter((s) => s.admin && s.enabledTarget).length > 0;
+
+    // Determine whether a scope that requires admin consent has been selected. An admin consent scope has been selected if isDirty = true.
+    let isDirty = PermissionScopes.filter((s) => s.admin && s.enabledTarget).length > 0;
+
+    // Reduce the size of the scopes table list by the size of the message bar. We only want to make this
+    // change the first time that the selected scope list is changed. 
+    if (isDirty && !this.hasRequestedAdminConsent) {
+      // Convert the table height from string to number.
+      var currentHeight: number = Number(this.scopesListTableHeight.replace(/px/, "")).valueOf();
+
+      // Update the table height based on the height of the message bar that's displayed when an admin consent scope is selected.
+      // This should keep the buttons from moving below the viewport as long as the component is showing in the viewport.
+      var updatedHeight: number = currentHeight - 135;
```

Comment:
Can we replace the "135" with the height of the message bar?

Reply:
That is the correct way to go about this. The message bar doesn't get added to the DOM until scopeListIsDirty = true which is too late. There must be a way to do this correctly, and I'm currently learning about the Angular lifecycle hooks which should give me ways to update the component after it has been added and preferably before, it has been rendered. Until I know how to update this based on component lifecycle, this fix will address the bug.

Reply:
👍